File: app/src/main/java/app/solocoin/solocoin/worker/LegalChecker.java
package app.solocoin.solocoin.worker;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.provider.Settings;
import android.util.Log;


import com.scottyab.rootbeer.RootBeer;

import java.util.List;
import java.util.Objects;

/*
* Created by Vijay Daita on 5/20
 */

public class LegalChecker {
    private Context context;

    public LegalChecker(Context context){
        this.context = context;
    }
    public boolean isCheating() {
        RootBeer rootBeer = new RootBeer(context);
        return rootBeer.isRootedWithBusyBoxCheck() || isMockSettingsON(context) || areThereMockPermissionApps(context);
    }

    private static boolean isMockSettingsON(Context context) {
        // returns true if mock location enabled, false if not enabled.
        return !Settings.Secure.getString(context.getContentResolver(),
                Settings.Secure.ALLOW_MOCK_LOCATION).equals("0");
    }

    private static boolean areThereMockPermissionApps(Context context) {
        int count = 0;

        PackageManager pm = context.getPackageManager();
        List<ApplicationInfo> packages =
                pm.getInstalledApplications(PackageManager.GET_META_DATA);

        for (ApplicationInfo applicationInfo : packages) {
            try {
                PackageInfo packageInfo = pm.getPackageInfo(applicationInfo.packageName,
                        PackageManager.GET_PERMISSIONS);

                // Get Permissions
                String[] requestedPermissions = packageInfo.requestedPermissions;

                if (requestedPermissions != null) {
                    for (String requestedPermission : requestedPermissions) {
                        if (requestedPermission
                                .equals("android.permission.ACCESS_MOCK_LOCATION")
                                && !applicationInfo.packageName.equals(context.getPackageName())) {
                            count++;
                        }
                    }
                }
            } catch (PackageManager.NameNotFoundException e) {
                Log.e("Got exception ", Objects.requireNonNull(e.getMessage()));
            }
        }

        return count > 0;
    }
}

File: app/src/main/java/app/solocoin/solocoin/worker/LegalChecker.kt
package app.solocoin.solocoin.worker


import android.content.Context
import android.content.pm.PackageManager
import android.provider.Settings
import android.util.Log
import app.solocoin.solocoin.app.SolocoinApp.Companion.sharedPrefs
import com.scottyab.rootbeer.RootBeer
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.InternalCoroutinesApi

/*
 * Created by Vijay Daita on 5/20
 */

class LegalChecker(var context: Context) {
    @InternalCoroutinesApi
    @ExperimentalCoroutinesApi
    fun isCheating(): Boolean {
        val rootBeer = RootBeer(context)
        var mock = false
        sharedPrefs?.let {
            mock = it.mock
        }
        return rootBeer.isRooted && mock
    }

    fun isMockSettingsON(context: Context): Boolean {
        // returns true if mock location enabled, false if not enabled.
        return Settings.Secure.getString(
            context.contentResolver,
            Settings.Secure.ALLOW_MOCK_LOCATION
        ) != "0"
    }

    fun areThereMockPermissionApps(context: Context): Boolean {
        var count = 0
        val pm = context.packageManager
        val packages =
            pm.getInstalledApplications(PackageManager.GET_META_DATA)
        for (applicationInfo in packages) {
            try {
                val packageInfo = pm.getPackageInfo(
                    applicationInfo.packageName,
                    PackageManager.GET_PERMISSIONS
                )

                // Get Permissions
                val requestedPermissions =
                    packageInfo.requestedPermissions
                if (requestedPermissions != null) {
                    for (i in requestedPermissions.indices) {
                        if ((requestedPermissions[i]
                                    == "android.permission.ACCESS_MOCK_LOCATION") && applicationInfo.packageName != context.packageName
                        ) {
                            count++
                        }
                    }
                }
            } catch (e: PackageManager.NameNotFoundException) {
                Log.e("Got exception ", e.message)
            }
        }
        return count > 0
    }
}
